crates/astria-sequencer/src/sequencer.rs

```diff
@@ -303,3 +286,104 @@ fn spawn_signal_handler() -> SignalReceiver {
         stop_rx,
     }
 }
+
+/// Returns a new Connect oracle client or `Ok(None)` if `config.no_oracle` is true.
+///
+/// If `config.no_oracle` is false, returns `Ok(Some(...))` as soon as a successful response is
+/// received from the oracle sidecar, or returns `Err` after a fixed number of failed re-attempts
+/// (roughly equivalent to 5 minutes total).
+#[instrument(skip_all, err)]
+async fn new_oracle_client(config: &Config) -> Result<Option<OracleClient<Channel>>> {
+    if config.no_oracle {
+        return Ok(None);
+    }
+    let uri: Uri = config
+        .oracle_grpc_addr
+        .parse()
+        .context("failed parsing oracle grpc address as Uri")?;
+    let endpoint = Endpoint::from(uri.clone()).timeout(Duration::from_millis(
+        config.oracle_client_timeout_milliseconds,
+    ));
+
+    let retry_config = tryhard::RetryFutureConfig::new(MAX_RETRIES_TO_CONNECT_TO_ORACLE_SIDECAR)
+        .exponential_backoff(Duration::from_millis(100))
+        .max_delay(Duration::from_secs(10))
+        .on_retry(
+            |attempt, next_delay: Option<Duration>, error: &eyre::Report| {
+                let wait_duration = next_delay
+                    .map(humantime::format_duration)
+                    .map(tracing::field::display);
+                warn!(
+                    error = error.as_ref() as &dyn std::error::Error,
+                    attempt,
+                    wait_duration,
+                    "failed to query oracle sidecar; retrying after backoff",
+                );
+                async {}
+            },
+        );
+
+    let client = tryhard::retry_fn(|| connect_to_oracle(&endpoint, &uri))
+        .with_config(retry_config)
+        .await
+        .wrap_err_with(|| {
+            format!(
+                "failed to query oracle sidecar after {MAX_RETRIES_TO_CONNECT_TO_ORACLE_SIDECAR} \
+                 retries; giving up"
+            )
+        })?;
+    Ok(Some(client))
+}
+
+#[instrument(skip_all, err(level = tracing::Level::WARN))]
+async fn connect_to_oracle(endpoint: &Endpoint, uri: &Uri) -> Result<OracleClient<Channel>> {
+    let mut oracle_client = OracleClient::new(
+        endpoint
+            .connect()
+            .await
+            .wrap_err("failed to connect to oracle sidecar")?,
+    );
+    let _ = oracle_client
+        .prices(QueryPricesRequest::default())
+        .await
+        .wrap_err("oracle sidecar responded with error to query for prices");
```

Comment:
i think we should move this out of the retried function and call it after we successfully connect to the client, as if the grpc call for prices fails, it indicates a more serious error that probably isn't fixable upon retry. thoughts?

Reply:
I think this is maybe a general problem with most or even all of our retry loops like this.  We probably have some error cases which are permanent/unrecoverable, and others which are transient.  However we always treat them as transient and keep retrying.

In this specific case, I guess we could also get a transient failure of the gRPC (e.g. timeout or dropped connection)?  We should probably look to abandon the retry loop if the error indicates a permanent failure, but keep trying if not.  However, that's another can of worms, since matching on error variants isn't a lot of fun when using eyre/anyhow.

Reply:
yeah, maybe we can do a larger revisit of the retry functions. this is fine for now, but something to keep in mind